01-algorithms/01-algorithms.md

```diff
@@ -50,65 +61,454 @@ The study of computer science is the study of algorithms.  Researchers in CS foc
+
+For our purposes we will expect the function to return an empty list `False` if the length of the list is less than two or if `k` is zero. In an interview setting, consider asking your interviewer if there is a preferred return value when the algorithm encounteres the edge case scenearios. If there is no preferred return value, make a thoughtful decision about what the algorithm should return and communicate your choice to your interviewer.
+
+<!-- available callout types: info, success, warning, danger, secondary, star  -->
+### !callout-star
+
+## Devising and Testing Go Hand in Hand
+
+These sample inputs and outputs make **excellent** test cases to use when verifying the algorithm we devise. Creating tests for an algorithm is often an integral step in helping us devise our algorithm!
+
+### !end-callout
+
+### Break Down the Problem
+
+Big massive problems are *hard*.  As developers it is often easier to break down the problem into smaller, easier to understand steps.  This also helps us make an application more modular, testable and maintainable.  Clearly not all problems need to be broken up, but most substantial problems in interviews and in the workplace do.
+
+How could you break up our sample problem?  Answer and then check the explanation below.
+
+### !challenge
+
+* type: paragraph
+* id: 819220a4-080c-4818-a984-29766400d819
+* title: What subproblems do you see?
+
+##### !question
+
+How would you break down the problem into two or more subproblems?
+
+##### !end-question
+
+##### !placeholder
+
+##### !end-placeholder
+##### !hint
+When you think about attempting to solve this problem, what about it seems difficult to you? Where are you getting stuck? 
+
+Feeling stuck often means you've found a subproblem you need to solve!
+##### !end-hint
+
+##### !explanation
+
+Sample subproblems could be:
+
+1.  Finding all sublists of size `k`.
+2.  Determining if a single sublist has duplicate elements.
+
+##### !end-explanation
+
+### !end-challenge
+
+### Solve or Reduce the Problem
+
+Once a problem is understood and broken into more manageable pieces we can then solve the problem.
+
+Write your solution and then look below to see a sample solution.
+
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: 4dc81336-1c2d-421e-8fe8-27326736b2e3
+* title: Duplicates with K Problem
+* points: 1
+* topics: python, lists, hashtables
+
+##### !question
+
+Write a function called `duplicates_within_k` that takes in a list of integers numbers, `numbers`, and an integer `k`.
+
+The function should return `True` if the same integer occurs multiple times within `k` distance. It should return `False` otherwise.
+
+Spend no more than 15 minutes working through this independently. Use the hints below or reach out for help if you are still feeling stuck after 5 minutes.
```

Comment:
Does this mean we want them to work on it for 15 minutes then reach out if they feel stuck for more than 5 minutes?

Reply:
Should be 15 minutes for both, thanks for catching that!